### ISSUE TITLE
Remove restrictions on netCDF object names

### ISSUE THREAD

**Issue by Dave-Allured** [cf-convention/cf-conventions#237]
**Title:**  Remove restrictions on netCDF object names

**Moderator:**

**Moderator Status Review:**  New issue, 2020 January 23

**Requirement Summary:**  None.

**Technical Proposal Summary:**  Remove [CF 1.7 section 2.3](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_naming_conventions) restrictions on characters in names of variables, attributes, etc.  Resolve ambiguous use of such restrictions.

**Benefits**
* Support international usage.
* Allow special characters in names.
* Remove ambiguity over requirement versus preference.
* Simplify CF rules.
* Simplify conformance checking.
* Improve compliance for some existing data sets.

**Caveats**
* Breaks compliance with COARDS name rules, but is a superset of them.
* Some existing softwares can not handle non-traditional characters.  They would need upgrades, but only when presented with new files using expanded character set.

**Status Quo:**  Object names are now restricted to a traditional yet limited character set which does not accommodate many non-western languages, nor other desired naming patterns.

**Detailed Proposal:**  Change the first paragraph of [2.3 Naming Conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_naming_conventions) as follows.  The remainder of **2.3** is left unchanged.

**Current version (1.8 draft):**
* Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores. Note that this is in conformance with the COARDS conventions, but is more restrictive than the netCDF interface which allows use of the hyphen character. The netCDF interface also allows leading underscores in names, but the NUG states that this is reserved for system use.

**Proposed:**
* Variable, dimension, attribute, and group names are not generally restricted by this convention.  Any names that are acceptable to the netCDF library may be used.  The most notable rules from netCDF are ASCII or UTF-8 character set, forward slash "/" not allowed, and names should not begin with underscore or certain other special characters.  Refer to file format specs in the NUG for more details.

(Edit: Added forward slash "/" after following comments were posted.)

**Comment by JimBiardCics** [comment]
While I generally approve of relaxing the character set restrictions, I think we may need to consider certain patterns that should either be reserved or restricted. As an example, the use of slashes ('/') in names wreaks havoc with group path formalisms that are already in place outside of CF. In addition to the prohibition on having leading underscores that is mentioned in the proposal, the netCDF-LD project (@marqh) is making use of doubled underscores within a name as a mechanism for marking namespaces. There may be other cases "in the wild" where certain patterns are in use, and I think we should be careful to avoid causing problems by being overly loose here.

I suggest that, at minimum, we should disallow the use of slashes ('/') or backslashes ('\') in names, and should call out two or more sequential underscores ('__') as reserved.

**Comment by steingod** [comment]
I support the constraint indicated above. Especially allowing slashes and backslashes in names will be confusing.

**Comment by erget** [comment]
Agreed, I think it would be best if the restrictions were presented in a table for readability.

**Comment by marqh** [comment]
We may get some benefit form considering other standardisation activity in this domain?  

RFC3986 defines the generic syntax for the Universal Resource Identifier (URI)
https://tools.ietf.org/html/rfc3986

As netCDF variables are resources that are being identified within the domain of a netCDF file, could we benefit from just adopting RFC3986?

This has a reserved character section:
https://tools.ietf.org/html/rfc3986#section-2.2

Disclaimer: I have not cross referenced this in detail with the NUG to examine consistency or problem areas (potential for contribution if useful)
First glance, these look pretty similar.

If these are consistent, then adopting the NUG definition unchanged looks sensible to me. It already mandates against the use of a '/' character, which is the most problematic one for me, given groups and variable identity within groups.

I'd like to see an explicit reference to the relevant NUG section in the text or linked, as I had to search a bit and I know what I'm looking for
I think:
https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_set_components.html#Permitted
is stable enough for a standards document 
(@ethanrd do you agree this is a stable URI for the resource please?)

mark

**Comment by JimBiardCics** [comment]
@marqh I like the overall suggestion of RFC3986. I think we should not adopt the "% encoding" concept of RFC3986. And, again, I think we should reserve leading "_" characters (per NUG) and multiple sequential "_" characters (per netCDF-LD). Are there any other special character sequences in the wild that anyone is aware of — in UGRID or Radial perhaps?

I notice that the NUG section you referenced implies that space characters are allowed as long as they are not at the end of a variable name. Do we want to allow internal spaces?

**Comment by marqh** [comment]
> @marqh I like the overall suggestion of RFC3986. I think we should not adopt the "% encoding" concept of RFC3986. And, again, I think we should reserve leading "_" characters (per NUG) and multiple sequential "_" characters (per netCDF-LD). Are there any other special character sequences in the wild that anyone is aware of — in UGRID or Radial perhaps?

I agree, @JimBiardCics, that adoption of %encoding is not a path I would want to walk.  it's perhaps a useful cross reference, but points like this suggest against including some specific use of RFC3986 within CF

> I notice that the NUG section you referenced implies that space characters are allowed as long as they are not at the end of a variable name. Do we want to allow internal spaces?

internal spaces!?!? really

if we can stop that, then that is a good thing.  Why would the NUG allow variable names with spaces in them??

my reading of 
> The names of dimensions, variables and attributes (and, in netCDF-4 files, groups, user-defined types, compound member names, and enumeration symbols) consist of arbitrary sequences of alphanumeric characters, underscore '_', period '.', plus '+', hyphen '-', or at sign '@', but beginning with an alphanumeric character or underscore. However names commencing with underscore are reserved for system use.

lead me to view space as not allowed.  However the following:

> Beginning with versions 3.6.3 and 4.0, names may also include UTF-8 encoded Unicode characters as well as other special characters, except for the character '/', which may not appear in a name.
> Names that have trailing space characters are also not permitted.

Could someone from a Unidata background confirm or deny that in netCDF4, a space may be used within a variable name?

**Comment by zklaus** [comment]
I have zero Unidata authority, but I'd like to state the obvious: Unicode is complicated.
This may already account for the somewhat vague formulation in the NUG if one takes a look at [the list of whitespace characters in unicode](https://en.wikipedia.org/wiki/Unicode_character_property#Whitespace). Indeed, whether one wants to go with a blacklist or a whitelist approach, it may be a good idea to think and write in terms of Unicode character categories (cf [here](https://en.wikipedia.org/wiki/Unicode_character_property#General_Category) or [here](https://unicodebook.readthedocs.io/unicode.html#categories)).

**Comment by ngalbraith** [comment]
I'm afraid I'm the odd man out here - I don't think the list of benefits in the original issue stacks up against the costs; in fact some of them don't seem to BE benefits. Maybe some use cases would be helpful ... Could you elaborate on how this change would support international usage? 

Is improved compliance for some existing data sets really a goal? What's in these data sets that needs to be described with a name that begins with a number or contains spaces or special characters?   

Maybe this is a selfish concern - we use Matlab's built-in netCDF library, and I'm not sure how that would deal with this change. If it's really needed for some specific reason, we'll deal with it, but absent that explanation, this is just a headache for a lot of CF users.

**Comment by ethanrd** [comment]
Is there a user asking for this extension, a particular use case that needs addressing? CF has generally tried to avoid extensions that seem like a good idea but don’t have a current use case.

Having said that, if we do move forward, I think we should be very cautious. Not only is Unicode very complicated as @zklaus points out, so are the rules around reserved character sets in URLs (and in which part of the URL) and file systems. Extending the set of characters allowed to include those reserved characters means they will need to be properly encoded when used in URLs (e.g., OPeNDAP and OGC WCS). Which, it turns out, isn’t as easy as it might seem.

Also, this or similar proposals/discussions have come up before, I think several times but so far I've only found these two:

- A 2014 discussion on the email list (the initial email is [here](http://mailman.cgd.ucar.edu/pipermail/cf-metadata/2014/006929.html)) focused mainly on expanding the set of characters allowed to include ‘@’, ‘+’, ‘-’, and ‘.’ with some mention of Unicode coming fairly late in the discussion. 
- Trac Ticket #[157](https://cf-trac.llnl.gov/trac/ticket/157) suggested moving from “should” to “must” on the current set of allowed characters.

**Comment by ethanrd** [comment]
@WardF and @lesserwhirls - Could you address the question of whether whitespace characters are allowed in netCDF variable names?

**Comment by MTG-Formats** [comment]
Having blank spaces in names would break other CF conventions like use of the ancillary variables attribute.

"The attribute ancillary_variables is used to express these types of relationships. It is a string attribute whose value is a blank separated list of variable names. "

How to parse this?
  float q_error_limit(time)
    q_error_limit:standard_name = "specific humidity standard error" ;
    q_error_limit:units = "g/g" ;

**Comment by taylor13** [comment]
I must be missing something, but if a variable is named, for example, "a-b", and one uses that in a computer code, how is it interpreted?  How is that variable distinguished from the operation: subtract variable "b" from variable "a"?  Don't "+", "-", "/", "*", " " all have this problem?

**Comment by JimBiardCics** [comment]
@taylor13 Your code would have to parse the variable name into code. Until you did something like that, it is just a string.

**Comment by taylor13** [comment]
As a user of data, I usually like the names of my variables (in my codes) to be the same as their names in the netCDF file.  With the current naming convention for CF, this is always possible, I think.  If, however certain restrictions were removed, as suggested above, this would no longer be true.  
I would echo others and ask what particular use cases are driving this?

**Comment by Dave-Allured** [comment]
Well, thank you for all yout thoughtful responses.  I see that we are rehashing [the 2014 discussion](http://mailman.cgd.ucar.edu/pipermail/cf-metadata/2014/006929.html), and probably others.  Thanks @ethanrd for finding that.  There are good arguments pro and con there, and it is worth reading.

The difference is that only 4 extra characters were proposed in 2014.  I simply want to legalize all the other 137 thousand!

> Is there a user asking for this extension, a particular use case that needs addressing? CF has generally tried to avoid extensions that seem like a good idea but don’t have a current use case.

No, I do not have a current use case.  This is a recurring issue, so I thought this comprehensive approach would be beneficial.  Past use cases were mentioned or implied in the 2014 discussion, and in [trac 157](https://cf-trac.llnl.gov/trac/ticket/157).

NetCDF developers put some care into expanded name capability, 12 years ago.  However, CF restrictions are copied virtually unchanged from 25 year old COARDS rules, which were probably based on ASCII only.  CF is overdue to allow the full naming range for creative purposes by all scientific users.

Name quoting is generally easy and well supported in most modern programming languages.  This takes care of UTF-8, math symbols, and other active characters.  IMO, naming freedom should outweigh exactly matching names of program variables.

**Comment by ngalbraith** [comment]
> @taylor13 Your code would have to parse the variable name into code. Until you did something like that, it is just a string.

Not everyone writes their own netCDF translators, and some packages no doubt take the variable and attribute names from the netCDF variable and attribute names. Those who use these packages are least likely to be in a position to accommodate this change.

When I have a minute I'll give it a try with the Matlab netCDF interface.  I'd be much happier to spend the time on it if there was more than 'creative purposes' for a reason. The [trac ticket](https://cf-trac.llnl.gov/trac/ticket/157) has an example of isotopes with names that begin with a number, which has some weight, but the work around for that seems simple compared to what would be needed by someone using code that auto-assigns variable names.

On the other hand, most folks probably work with multiple standards; OceanSITES would no doubt maintain the variable name restriction, if CF doesn't.

**Comment by zklaus** [comment]
I agree that it would be good to have use cases.

@ngalbraith is also right that not everyone is writing their CF code based on naked netCDF access. Indeed, I consider such an approach foolish, since CF is far too rich by now to stand a series chance of getting it right.

However, while using the netCDF variable name as a program variable name might be excused in small, not reused code that only ever will deal with, say `tas`, it is inexcusable in general-purpose library code. How would such a variable enter the namespace without the program knowing its name beforehand? Ultimately, the only way is via the equivalent of `eval(var_name)`. Such code is prone to breakage no matter what restrictions we put on the character set since it would always leave open the possibility of having reserved words of the particular programming language as variable names. Another serious problem is that it opens the possibility to maliciously crafted variable names: How about `var_name='system("rm -rf .")'`?

Hence, I don't think the argument that all netCDF variable names should be permissible program variable names in all programming languages should guide the design of CF.

**Comment by DocOtak** [comment]
I had the same thoughts as @zklaus when thinking about the security implications of what I could only imagine was an `eval(var_name)`. I've even seen some of the matlab code which does exactly this to load all the variable into a matlab namespace. I'd even go so far as to recommend that the CF document itself warn against doing this...

**Comment by martinjuckes** [comment]
I agree that some use cases would be helpful. I'm not sure about the specific proposal that initiated the discussion, but I do agree with the thought behind it that we should have a considered and reasoned policy on this, rather than just having a frozen-in rule based on past library constraints. 

One reason that we might want to depart from the full freedom allowed in NetCDF is that we have, in CF, a range of different attributes to describe a variable. The `long_name` is designed to hold human readable text, the `standard_name` and `units` which both have strongly constrained values. 

Some application libraries need, in places, identifiers with a restricted character set. For example, I can construct a `collections.namedtuple` with name `tas`, but not with name `tas.Amon` because, in python "Type names and field names can only contain alphanumeric characters and underscores" (cited from an error message generated by `collections.namedtuple`). Could this be considered as a use case for having place in the convention to specify, for CF objects, an identifier which is composed of "alphanumeric characters and underscores"? The variable name is the de facto place which many people use for this kind of identifier (perhaps because of legacy packages). 

Note that the `standard_name` fits the character restriction, but does not fit the use case because different variables may have the same `standard_name`.

Another potential use case is for identifiers of concepts described in [RDF Turtle](https://www.w3.org/TR/turtle/) which has a character restriction on object names, broader, I think, than  "alphanumeric characters and underscores", but definitely narrower than 137 thousand available of UTF-8.

The desire to have a simple identifier is linked, in my mind at least, to the concept of a namespace, which is being discussed in the context of NetCDF (see [NetCDF-ld](https://github.com/opengeospatial/netcdf-ld) and discussion on [namespace delimiters](https://github.com/opengeospatial/netcdf-ld/issues/50)). I don't this is simply a matter of upgrading software to make it accept generic strings: there is a wide range of applications that exploit identifiers constructed from a limited character set in order to enable the use of identifiers within an text string.

**Comment by zklaus** [comment]
One potential use-case that always came to my mind without an actual example at hand Is the native names of weather stations, say a temperature time-series from the Umeå station, where the variable name contains the station name.

What makes this particularly interesting is that it seems to be permitted already under current CF conventions, since under [CF-1.8, Section 2.3 Naming Conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#_naming_conventions) it says:

> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores. [...] Languages other than English are permitted for variables, dimensions, and non-standardized attributes.

**Comment by martinjuckes** [comment]
HI @zklaus : good point about the existing rules.

Regarding your use case; wouldn't that use case be covered by setting the `long_name` to "Temperature time-series from the Umeå station"? The current convention appears to permit "Umeå_station", but not "Umeå station" (blanks not allowed).

The cfchecker (4.0) takes a narrower view of what is allowed, restricting variable names to string matching the python regex: `'^[a-zA-Z][a-zA-Z0-9_]*$'`.

**Comment by zklaus** [comment]
Yes, that might be a good way to encode the information. What I wanted to say is this: I find it very plausible that in a national weather service a group sits together and decides to code their station data using variable names `tas_station-name` with a number of non ascii letters in the station names. Furthermore, that would appear to be perfectly valid CF.

So I think being more explicit about what is meant by "letter" would be good, even if that means saying that only ascii letters are allowed.

**Comment by sval-dev** [comment]
In case it is helpful to have a real use case, the desire to have variables and groups able to describe "PM2.5" is described in the related discussion at https://github.com/cf-convention/discuss/issues/256 and an example prototype product making use of these group and variable names can be found at:
https://asdc.larc.nasa.gov/data/MAIA/L4_GFPM_VSIM001/2018/01/MAIA_L4_GFPM_20180101T000000Z_FB_NOM_R01_USA-Boston_F01_VSIM01p01p01p01.nc

**Comment by ethanrd** [comment]
Hi all - Just caught up on the conversation in #256. I wanted to mention that the Zarr specification group had a related discussion last January or so (Issue #[56](https://github.com/zarr-developers/zarr-specs/issues/56) "Node name character set" and PR #[196](https://github.com/zarr-developers/zarr-specs/pull/196)). It was pretty focused on expanding to include non-ASCII Unicode characters. There was some good discussion on Unicode normalization and how to restrict the set of allowed characters. At this time, a recommendation for the Unicode normalization was added to the Zarr v3 core specification (see "Node names" [section](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#node-names)) but any recommendation for a restricted set of allowed characters was put off to an extension. (The Zarr v3 specification is being developed with a core and extensions and conventions model.)

Along with the Python Langague [syntax for identifiers/names](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) (mentioned in issue #256), the Zarr discussion also include the following references that might be useful:
- Python [PEP 3131](https://peps.python.org/pep-3131/) "Supporting Non-ASCII Identifiers"
- Unicode General Category property of each character (code point) - ([Wikipedia page](https://en.wikipedia.org/wiki/Unicode#General_Category_property))
- [Unicode Normalization FAQ](https://unicode.org/faq/normalization.html)
- [Unicode Standard Annex #31](https://www.unicode.org/reports/tr31/): Unicode Identifier and Pattern Syntax
- [Unicode Technical Report #36](https://www.unicode.org/reports/tr36/): Unicode Security Considerations

**Comment by ethanrd** [comment]
My main take away from the Zarr discussion is that the Python Language [syntax for identifiers/names](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) (mentioned in issue #256) seems like a good starting point (assuming we want to open things up as wide as possible, the other 137K characters as @Dave-Allured said). For maximum interoperability while supporting Unicode, I think further restrictions would be necessary since netCDF variable names often end up in URLs and file system names.

While encoding/escaping of characters/bytes (e.g., url-encoding) is well supported, when to encode, with which encoding, and by what software component can get confusing. So I don't think character encodings fully address interoperability affects of reserved characters in URLs and file system names.

**Comment by larsbarring** [comment]
If this issue is now revived, it might be a good time to find a moderator (I am not volunteering because of limited expertise and experience).

**Comment by sethmcg** [comment]
I think that any form of whitespace should be disallowed in naming.

In addition to the aforementioned problem it causes with ancillary variables, it's not uncommon in my experience for a lot of netcdf processing to happen on the command line (rather than neatly encapsulated within the confines of a general-purpose library) by piping the output of `ncdump -h` through various commands, in particular `cut` and `grep`.  Whitespace would play havoc with those kinds of workflows, which in my opinion makes it an absolute showstopper.

This use case also makes me very leery about the prospect of allowing any character that is a special character in the shell or a regular expression.  Let's not set up a situation that demands lots of quoting.

Further, this suggests to me that if the list of allowed characters is to be expanded, it should be via a whitelist approach rather than a blacklist approach; i.e., the default should be that characters are disallowed unless they have been carefully vetted.

**Comment by JonathanGregory** [comment]
I agree with all of @sethmcg's points.

**Comment by larsbarring** [comment]
This is just to record a use case by @markusfiebig in [discuss/#256](https://github.com/cf-convention/discuss/issues/256#issuecomment-1671061216) (that I am going to close as duplicate):

> I would in fact propose to relax the character restrictions for CF names considerably since these restrictions limit the usability of the convention. I will soon have to propose names for the concentrations of PCBs, so we are looking at names of the type
>
> 2,2',3,3',4,4',5,5'-octachlorobiphenyl mass concentration
> 
> The commas and quotation marks in this name are essential to denote the chemical, so they can't be replaced. PCBs and brominated flame retardants are clearly a relevant area of atmospheric research and need to have a place in the CF naming convention.
>
> To limit the character set of a vocabulary to meet the needs of programming languages is rather outdated. Programming languages should serve the use cases, not limit them.

**Comment by Dave-Allured** [comment]
Regarding program name spaces, quoting, whitespace, etc; it is the programmer's responsibility to be alert to special cases, avoid namespace conflict and code injection, and use quoting as needed.  The role of CF should be to describe metadata, not to guard against names reasonably crafted from the rules of the underlying format.

Let's add something to deal with CF's blank-separated lists.

* Variable names included in blank-separated lists such as `ancillary_variables` or `coordinates` must not include the ASCII space character.

**Comment by ethanrd** [comment]
I think CF should make recommendations regarding situations that may hinder interoperability. Allowing Unicode characters in variable names will add many challenges to interoperability.

The current language in section 2.3 says "names should begin with a letter and be composed of letters, digits, and underscores". In ASCII that translates to the `'^[a-zA-Z][a-zA-Z0-9_]*$'` mentioned by @martinjuckes above. I believe the Python Language [syntax for identifiers/names](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) I mentioned above is close to the same for Unicode, the underscore gets expanded to the Unicode Connector Punctuation (Pc) category and some mark categories as well.

I think each of those restrictions would be good to mention in CF as providing two different levels of interoperability.

**Comment by ethanrd** [comment]
The 2023 CF Workshop is next week, 3-5 Oct 2023 (agenda and registration info [here](https://cfconventions.org/Meetings/2023-Workshop.html)). I have been thinking of trying to present an overview of some of the Unicode issues involved in this discussion as a lightning talk. If enough folks attend the workshop interested in further discussion on this topic, we could spin up a hackathon breakout.

**Comment by Dave-Allured** [comment]
> I think CF should make recommendations regarding situations that may hinder interoperability.

Naming interoperability should be governed by the underlying file format, not CF.

**Comment by larsbarring** [comment]
I must say that the more I think about this issue the more uncertain I get about the intention and scope of the suggested changes. It might very well be that I am confusing myself or misunderstanding the intent, but here is my current thinking summarised into a few points:
* The text in section 2.3 specifying the CF naming requirements only applies to what constitutes CF metadata, not *everything* in a netCDF file. 
* The netCDF file format accepts almost all unicode characters in names, so in that sense there is no restriction on creativity.
* Just as a random example take the CF attribute `long_name`. I might want to use the Swedish language version `långt namn`, someone else might want to use the Amharic one `ረጅም ጊዜ` (from an online translation site), and so forth. There might be perfectly reasonable arguments for doing that at national levels, but does it help interoperability? Is this something that CF should cater for?
* From the perspective of CF I have a hard time understanding how naming freedom would go hand in hand with interoperability. In a sense standards and conventions are always restricting "creativity"(*) just to create interoperability. Just think of USB connectors, or wall plugs when travelling. For these two examples I am happy with all the standardisation work that resulted in rather few variants, still I personally think that anything more than just one alternative is too many. 
* Currently there are use cases for expanding the set of allowed characters in variable names. 
* We need to clarify what is meant by the word **letter** in the CF text.

(*) I write "creativity" because the restrictions equally much *creates* creativity, which arises from the fact that data analysts can think creatively about analyses rather wrestling with data harmonisation.

**Comment by Dave-Allured** [comment]
> The text in section 2.3 specifying the CF naming requirements only applies to what constitutes CF metadata, not _everything_ in a netCDF file.

@larsbarring, the opening paragraph of the existing section 2 clearly says "In this section we describe conventions associated with filenames and the basic components of a netCDF file".  I think your interpretation "only applies to what constitutes CF metadata" is mistaken.

This proposal applies to the general name space, such as the user's variable and attribute names.  It does not propose to change anything about any CF controlled vocabulary.

**Comment by larsbarring** [comment]
Ah, I now see what you mean --- thanks @Dave-Allured! Then I think that the core of the issue is how the text in section 2.3 interacts with the opening paragraph of section 2. It might be easier to change the opening paragraph to explain what is generally allowed in netCDF files and then state that the raining text deals with CF specifications. But all this will be discussed at the CF workshop.

**Comment by ethanrd** [comment]
Hi all - There will be time to discuss this issue tomorrow (4 Oct) during the CF Workshops hackathon breakout session, currently scheduled for 16:40 UTC (10:40 US Mtn). I gave a very brief summary today during the hackathon introductions session (slides 1-3 [here](https://docs.google.com/presentation/d/19qWmYFRPtQArJbmvVr78_98gI8xXGGVF9HtMFB_iUjk/edit)).
Hope you can join us for the discussion.

**Comment by Dave-Allured** [comment]
I will add **title.fr-CA** from the localization discussion, which in my opinion is the optimal solution.  There have also been other suggestions on that issue, involving other special characters.

**Comment by ethanrd** [comment]
@Dave-Allured - Thanks. I have added another slide listing both these use cases.

There was a question from @JonathanGregory in issue cf-convention/discuss#256 about whether the intent for the octachlorobiphenyl use case was for variable names or for standard names. Do you know if that got answered? I'm not finding it in there.

**Comment by Dave-Allured** [comment]
Ethan, I do not know.  I refer you to the original author, @markusfiebig.

**Comment by larsbarring** [comment]
I will be in another breakout group, so here is a tougth: Might it be useful to have different rulees for attribute names and variable names?

**Comment by ethanrd** [comment]
@Dave-Allured wrote:
> Naming interoperability should be governed by the underlying file format, not CF.

While I agree much of this should probably be more carefully thought out at the netCDF level, there is I think a need to mention all this in CF. Perhaps it is more a backwards compatibility issue than interoperability, both for software and for new data.

**Comment by turnbullerin** [comment]
> I think CF should make recommendations regarding situations that may hinder interoperability. Allowing Unicode characters in variable names will add many challenges to interoperability.
> 
> The current language in section 2.3 says "names should begin with a letter and be composed of letters, digits, and underscores". In ASCII that translates to the `'^[a-zA-Z][a-zA-Z0-9_]*$'` mentioned by @martinjuckes above. I believe the Python Language [syntax for identifiers/names](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) I mentioned above is close to the same for Unicode, the underscore gets expanded to the Unicode Connector Punctuation (Pc) category and some mark categories as well.
> 
> I think each of those restrictions would be good to mention in CF as providing two different levels of interoperability.

Here's some notes on interoperability with other file formats and some common tools (restrictions are on variable names and, if supported, attribute names - values tend to be more flexible)

* DAP2 allows [0-9A-Za-z_!~*'"-] and other US-ASCII if URL-escaped; Special Characters: =<>!+-/\*~%.[]
* DAP4 UTF-8 characters (escaped if not US-ASCII); Special characters: /
* HDF5 UTF-8 supported
* ASCII, CSV, TSV are character-encoding dependent but all valid characters allowed (with proper escaping)
* KML depends on coding, <& and either ' or " must be escaped and non-printable control characters and compatibility characters are discouraged: https://www.w3.org/TR/xml/#NT-Char
* ESRI strongly recommends [A-Za-z0-9_-], explicitly not allowed: +*/!^%()[]{},~'":;><&|\=@#$

It seems the direction is towards full UTF-8 compatibility but not all tools/platforms are there yet. 

My take on this is, at the very least, we should not allow UTF-8 control characters (so 0x00 to 0x0F, 0x7F, and 0x80-0x9F, general category Cc) along with the two slashes and following the W3 guidelines in section 2.2 of the above reference is a good starting place for a disallow list of characters.

**Comment by MaartenSneepKNMI** [comment]
Keep in mind that the letter category in python is already very broad. The statement:

```python
from math import pi as π
```

is legal in current python. That also means that all issues in URL's caused by the use of characters that have the same shape but are actually different codepoints will be included here. Spot the difference: "o" or "ο". So as a warning: be careful what you wish for.

**Comment by larsbarring** [comment]
I think this conversation is spreading out in many directions, and I think that it is useful to disentangle these.

**1.** To my understanding this issue was initiated because the current CF text is inconsistent in relation to the recent development of the netCDF library. Because NetCDF  now accepts almost any Unicode character the conventions text would become clearer if the opening text of chapter 2 and the text in section 2.3 better distinguish between what is general netCDF requirements and what is  CF requirements. (See [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1742949246)  and [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1743058655) for background.)
**2.** We need to clarify what is meant by the word **letter** in section 2.3. (See [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-732238536) and [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-732258060) for background.)
**3.** There are requests to expand the set of allowed characters for variable names. (See [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1747131323) and the more extensive discussion in [discuss/#256](https://github.com/cf-convention/discuss/issues/256).)
**4.** To allow multilingual support via standard locale identifiers there is a specific request to expand the set of allowed characters for attributes. See [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1747145857) and the full discussion in [discuss/#244](https://github.com/cf-convention/discuss/issues/244).)

Merging these into a general conversation about how CF should handle Unicode would, I am pretty sure, be a complex and perhaps protracted effort with many aspects and diverse views. We could treat **1** and **2** simply as `defects` and just update the conventions text. But  in the light of **3** and **4** this is not constructive. Moving **3** forward would require a thorough discussion of how CF handles variable names and how they are used in various software. By comparison, moving **4** forward might be comparatively simpler as it involves a very specific request for additional characters in attribute names. As there is some urgency to implement multilingual support I suggest that we deal with this separated from the more general question of opening up for a wider set of characters in attribute names. Hence, I suggest that we proceed as follows with the aim to get the outcome into version 1.11:

* First we deal with **4** (I  can hardly see a more worthy use case than adding multilingual support to certain attributes). The suggested specific characters to be allowed in attribute names are **`.`** (period) and **`-`** (hyphen), see [here](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1747145857), or alternatives discussed in [discuss/#244](https://github.com/cf-convention/discuss/issues/244).
* Then we update the text in the opening paragraph of chapter 2 and in section 2.3.

After that, or in parallel, we can continue the the more general conversation on widening the character set allowed for variable names and what implications that may have.

**Comment by larsbarring** [comment]
Hm, having thought a bit more about all this, and in particular the implications of the diverse limitations imposed by different software and applications [listed](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1749052117) by @turnbullerin I retract much of my previous comment. Apologies for the confusion.

**Comment by turnbullerin** [comment]
@larsbarring  honestly, for bullet 4, I would classify it as more of a want than a need. There are solutions that don't require anything but `a-zA-Z_` in variable and attribute names. 

In general though, I think we are seeing that there is a lot of complexity once you move away from just the NetCDF file itself and into various processing environments and conversions. That said, the domains where NetCDF is being used might be the best ones to manage that complexity rather than us trying to manage it for them. Just because they can do a thing doesn't mean they must do that thing.

Three approaches come to mind:

1. We reject this and limit everything to the current characters. We miss some use cases but we simplify interoperability.
2. We allow all sensible characters and start using them in CF standards (e.g. for multilingual cases). We add some use cases but break backwards compatibility with some tools that are relying on that standard. I would add a warning for that in the documentation that using the expanded character set may cause issues.
3. We allow all sensible characters but restrict CF standard attributes to the current characters. We add some use cases and backwards compatibility is broken but might be more easily fixed since (a) NetCDF allows those characters and (b) they are non-standard so they can be ignored.

2 and 3 also require a decision on "sensible characters".

In effect, this decision is about the proper representation of technical things in variable/attribute names vs. interoperability. For me, interoperability wins over trying to support every conceivable use case for naming. A variable name isn't the only way to document what is in a NetCDF variable (e.g. PM_2_5 can have `PM_2_5:long_name = "PM 2.5";`). A variable name is a convenient short-hand for accessing the data, not necessarily a fully complete description of its contents (though it is ideal if it is fairly descriptive).

My personal thought is that approach 3 is the right one - expand the character set as far as reasonably possible, but with a big RECOMMENDED to use [A-Za-z0-9_] (not starting with underscore) for backwards compatibility and then not using them in standard attributes to maximize interoperability with other standards (looking at ESRI and MATLAB as the two clearest examples that support these restrictions).

**Comment by ChrisBarker-NOAA** [comment]
> accessing attributes that can't be represented with Python variable names is more complex
indeed --what most important (in my mind) is that it is *possible* -- that is, changing the requirements won't require a change in the netCDF4 library.

Personally, I think it was a design mistake to use Python attribute access for netcdf attributes -- it a confusing mingling of data and code, and leads to jsut these problems (also -- what if an actual Python attribute, like a method name, conflicts with a Variable name? 

So you can use get/setncattr for any netcf valid name.

Inshort, I think the saoifteate should name be making any assumptions about the form of the names.

However, it's the reality that lot of code does make these assumption (ERDAP, too, yes?) -- so we may be stuck -- at least not without a long transition time.

**Comment by ChrisBarker-NOAA** [comment]
> We need to clarify what is meant by the word letter in section 2.3.

How about:  `[A-Za-z0-9_]`

**Comment by larsbarring** [comment]
@ChrisBarker-NOAA I think you hit the nail :+1: when writing
> design mistake to use Python attribute access for netcdf attributes -- it a confusing mingling of data and code
> ...
> we may be stuck -- at least not without a long transition time

But it may be easy to see the mistake now with all the modern language capabilities and tools, an not least insights into software design principles. But some 20 years ago it was not so obvious [at all].

**Comment by larsbarring** [comment]
> >    We need to clarify what is meant by the word letter in section 2.3.

> How about: [A-Za-z0-9_]

This presupposes that the reader knows enough about regexp (and this dialect), otherwise this does not make sense. It is probably  more confusing than helpful, because the text specifically rules out the hyphen.

The specific (first) sentence in section 2.3 currently reads 
> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores. 

This could be changes to something like (bold indicates additions):

Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores. **With the word _letters_ we mean the standard ASCII letters uppercase  `A` to `Z` and lowercase `a` to  `z`.** 

Changes to the subsequent sentence is dealt with by @Dave-Allured's PR #465.

**Comment by PythonCHB** [comment]
> With the word letters we mean the standard ASCII letters uppercase A to Z and lowercase a to z.

+1 -- always good to be clear and explicit.

Does this mean we given up on expanding the allowable charactors jsut for now?

In short, I think there are two issues:

1) Using "punctuation" charactors like ., -, * ... that are used by computer languages as having special meaning.

2) Using non-englich charactors that are used in natural language ("letters") -- this could be accented latin charators, or totally different alphabets, etc -- I *think* Unicode has some defintions that could be used.

Changing (1) would require breakage with software that uses netcdf names as variablenames, etc.

Changing (2) may not -- I only know about Python, but it allows a wide variety of charactors that are defined as "letters" in Unicode. I would think non-english speakers would appreciate that. (being pathetically mono-lingual, I'll let others chime in)

So is (2) still on the table?

NOTE: with (2) there is the complication of "normalization" -- thanks to the beauty of human language, there are sometimes more than one way (more than one codepoint, or set of code points)  to represent what humans perceive as the same letter. Python uses NFCK normalization, though it has its issues[*]. IN any case, that would have to be decided if we wanted to go beyond ASCII.

Python spec:
https://docs.python.org/3/reference/lexical_analysis.html#identifiers

Discussion of the above:
https://www.asmeurer.com/python-unicode-variable-names/

[*] I'm forgetting the details a bit, but the two issues I see are:

a) someone may not WANT the normalization -- e.g. in math, we use, say, BlackBoardBold letters to have a different meaning than the regular one -- you might not want them normalized.
b) the normalization doesn't necessarily occur everywhere -- for instance, when source code is parsed, the names are normalized, but when you use `getattr()` and `setattr()`, I don't think they are, and they certainly aren't if you directly edit an object's `__dict__` -- this has real implications for CF / netcdf, as names will be transferred programmatically via such meta-programming techniques.

**Comment by larsbarring** [comment]
> Does this mean we given up on expanding the allowable charactors jsut for now?

I most certainly hope not :-) 
But at the same time, we should clear up the current text so that it is up to date.

**Comment by DocOtak** [comment]
@PythonCHB object identifiers in netCDF files are UTF-8 normalized with NFC, IIRC this is forced by the netCDF library. Even the stuff people call "ASCII" is really UTF-8 with the restricted code points (one of the miracles of UTF-8 is the ASCII backwards compatibility)[^1].

In the documentation I've written recently, I tend to do a formal code point definition followed by a more human readable clarification. For example, I might write the current restrictions as follows:
> Object names must only contain the following inclusive ranges of code points encoded as UTF-8: U+0041 to U+005A (upper case A to Z), U+0061 to U+007A (lower case a to z). Object names may include, but must not start with: U+0030 to U+0039 (digits 0 to 9), and U+005F (the lowline or underscore _).

If you want to include the `[A-Za-z0-9_]` it might be useful to explicitly call it a regular expression.

[^1]: I've haven't personally seen real 7-bit ASCII data in the wild, or I was unaware of what it was.

**Comment by ChrisBarker-NOAA** [comment]
>> Does this mean we given up on expanding the allowable charactors jsut for now?

> I most certainly hope not :-)
> But at the same time, we should clear up the current text so that it is up to date.

Absolutely.

But maybe it's time for a new issue for the next step, with a clear proposal?
(at this point, I think it would be to expand the definition of "letter"without restricting the constraints on other characters).

As a monolingual english speaker, I'm not in a position to make the case, but I imagine someone is :-)

**Comment by ChrisBarker-NOAA** [comment]
> If you want to include the [A-Za-z0-9_] it might be useful to explicitly call it a regular expression.

I don't want that in particular -- a short verbal definition is better. I'm not fluent with Regex, but find that particular one pretty readable -- better than "letters and numbers" anyway :-).

I don't think it's necessary to include the octal code points -- that's perhaps a bit too geeky for the audience of this document.

**Comment by larsbarring** [comment]
Regarding the interpretation of  "ASCII", the "standard uppercase A to Z" etc. there are a number of possible pitfalls and complications if one really want to nail it down.

Yes, "ASCII" is 7 bit encodings, but when going to 8 bits there are complications and alternatives, which do not impact the "A to Z or a to z as such, but the $ vs. ¤ vs. Ұ, or [\]{|} vs. ÅÖÄåöä, and so on.  But I think that if we really want to go into such detail most readers would not understand anyway because then we have to reference, "International alphabet no 5",   ISO and ITU standards and what not. I do not think that would be particularly helpful.

Moreover, if we want to be knitpicking and skip a superficial reference to the ASCII character set and just write "uppercase A to Z and lowercase a to z" then we end up in problems because in some language that range would include some combination of  û, ũ, è, é, á, à, ç, ñ .... (though Swedish is safe because we have the ÅÄÖ/åäö after the Z/z :-).

Anyway, this is why I thought that the sentence
_"With the word letters we mean the standard ASCII letters uppercase A to Z and lowercase a to z."_ 
is a reasonable approximation without digressing into tangential details. If anyone have a better suggestion please do feel free to comment.

**Comment by ChrisBarker-NOAA** [comment]
> "With the word letters we mean the standard ASCII letters uppercase A to Z and lowercase a to z."

Agreed -- precise but not too technical.

**Comment by larsbarring** [comment]
In order to make the current text (draft version of CF-1.11) up to date, and on the advice of @davidhassell, I will open a new issue to take care of the first few sentences of section 2.3 (cf. [this comment](https://github.com/cf-convention/cf-conventions/issues/237#issuecomment-1786748549)).  

I think that the excellent text suggestion in PR #465 would fit well in the new issue --- if that is OK with you  @Dave-Allured?

**Comment by Dave-Allured** [comment]
@larsbarring, it looks like your letters proposal targets a different paragraph, and it is not in conflict with PR #465.  Please allow #465 to proceed normally as an independent proposal.  Please submit a separate PR for your own proposal.

**Comment by JonathanGregory** [comment]
A small point: It should be "**By** _letter_ we mean"

**Comment by larsbarring** [comment]
@Dave-Allured If I am not mistaken it is the same paragraph and the very sentence immediately preceding the one you suggest to change. Here is a screenshot from viewing your PR (the section numbering is not rendered):

![image](https://github.com/cf-convention/cf-conventions/assets/15440426/cd376faa-79ae-4c90-b8d5-66c2f1cd1605)

Thanks @JonathanGregory  :+1:

**Comment by Dave-Allured** [comment]
@larsbarring, my bad.  You are correct, your changes are in the same paragraph as #465.  Please excuse my hasty reply.

Nonetheless, I prefer that you keep these two proposals separate.  It looks like your letters proposal is not in any way in conflict with #465. Please allow #465 to proceed normally as an independent proposal. Please submit a separate PR for your own changes.

**Comment by Dave-Allured** [comment]
I have started CF #477 to enable only two characters, period (.) and hyphen (-), in attribute names only.  This is a subset of my full proposal to remove restrictions on all netCDF object names.  CF #477 is in support of CF #244, localization of metadata (multilingual support).

Comments are invited.  If you agree with adding these two characters for attribute names only, please post a supporting comment on #477.

**Comment by JonathanGregory** [comment]
I believe that this issue is waiting for the outcome of #477 - is that right, @Dave-Allured?

**Comment by Dave-Allured** [comment]
@JonathanGregory, no, this issue is not waiting on #477.  This issue #237 is a free-standing proposal to remove all CF-specific restrictions on Netcdf object names.  In my view, this #237 is currently an open discussion, and waiting vaguely on a general consensus.

**Comment by ethanrd** [comment]
Hi Lars @larsbarring - I believe this [OGC netCDF-LD GH repo](https://github.com/opengeospatial/netcdf-ld) is the more current one. It provides a link to the [OGC netCDF-LD draft specification](https://docs.ogc.org/DRAFTS/19-002.html).

The OGC process involves a public comment period before proceeding to a vote. If I'm remembering correctly, the specification went out for public comment but hasn't yet gone out for a vote. Mark @marqh may be able to provide more details.

**Comment by sethmcg** [comment]
> In my view, this #237 is currently an open discussion, and waiting vaguely on a general consensus.

If this is waiting on general consensus to come to a resolution, I'll jump in and say that I oppose this proposal.

A lot of very serious interoperability and security concerns have been raised about the idea of removing all restrictions on naming, and I don't see any benefits that outweigh them.  Moreover, we don't have an actual motivating use case; this is an anticipatory change, which CF generally tries to avoid.

I'm open to motivated proposals that extend the allowed set of characters in a specific and more limited way, such as #477 (which has been accepted and is just waiting for a PR), but I think the discussion there demonstrates why it's important to be conservative and carefully discuss all the impacts of adding new allowed characters.

**Comment by larsbarring** [comment]
I fully agree with @sethmcg. 

Moreover, the opening sentence of [Section 2.3](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#_naming_conventions) reads
> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores. By the word letters we mean the standard ASCII letters uppercase `A` to `Z` and lowercase `a` to `z`. By the word digits we mean the standard ASCII digits 0 to 9, and similarly underscores means the standard ASCII underscore `_`. 

where the operative word is **should**, which, if [we interpret it](https://github.com/cf-convention/discuss/issues/258) as being in uppercase according to [BCP14/RFC2119 ](https://datatracker.ietf.org/doc/bcp14/) means:
> SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

This interpretation of "should" strikes me as a reasonable balance between strictness/limitations and openness/flexibility. If the CF Community moves to introduce BCP14 in the Conventions document there is of course the possibility that the word **should** is replaced by **MUST**, but that is a good time to revisit this issue.

**Comment by JonathanGregory** [comment]
The opening sentence of Section 2.3 states that

> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

Lars is correct that the word "should" here is a recommendation, as is clarified by Sect 2.3 of the [conformance document](https://cfconventions.org/cf-conventions/conformance.html). The conformance document further clarifies it

> This corresponds to ASCII characters in the decimal ranges (65-90), (97-122), (48-57), and (95). The corresponding Unicode codepoints are (U+0041-U+005A), (U+0061-U+007A), (U+0030-U+0039), and (U+005F).

and both the standard and the conformance document add (again, as a recommendation)

> ASCII period (.) and ASCII hyphen (-) may also be included in attribute names only.

which results from the agreed proposal https://github.com/cf-convention/cf-conventions/issues/477 of @Dave-Allured.

@larsbarring and @sethmcg have expressed views against a blanket removal of restrictions on the characters to be used in CF-netCDF object names. I agree that removing all restrictions would not be consistent with the usual CF approach. Normally, we consider specific proposals to change the status quo, motivated by present use cases. Are the other views on this question? It would be good to reach a consensus. Thanks.

**Comment by larsbarring** [comment]
The sections @JonathanGregory points at essentially provide *whitelist* of explicitly allowed characters, all other characters are not recommended (or recommended against) but not explicitly disallowed. But throughout this conversation there have been several remarks that some characters should indeed be explicitly disallowed. This could easily be done by amending the text in section 2.3 to list which character and character ranges CF explicitly disallows, i.e. creating a *blacklist*.  All other characters would then belong to a "*greylist*" where users are on their own and cannot expect the same level of interoperability and support from common libraries and software tools.

**Comment by Dave-Allured** [comment]
> the word "should" here is a recommendation, as is clarified by Sect 2.3 of the conformance document

This wording with "should" is confusing and unfriendly in context of that opening paragraph on netCDF object names.  Witness multiple tickets filed to remove character restrictions which did not really exist.  If that were simply reworded to clearly express the allowed versus recommended character sets, that would be sufficient.  CF is for scientists and programmers, not lawyers.

**Comment by JonathanGregory** [comment]
We've already agreed [elsewhere](https://github.com/cf-convention/discuss/issues/258) that we will check all the "must", "should" etc. words to make them conform to BCP-14, in which "should" indicates a recommendation. In this case, our interpretation has apparently changed. The text in sect 2.3

> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

has been the same since CF version 1.0. However, up to [version 1.7 of the conformance document](https://cfconventions.org/Data/cf-documents/requirements-recommendations/requirements-recommendations-1.7.html) this was listed as a _requirement_

> Variable, dimension and attribute names must begin with a letter and be composed of letters, digits, and underscores.

In [version 1.8 of the conformance document](https://cfconventions.org/Data/cf-documents/requirements-recommendations/conformance-1.8.html) it turned into a _recommendation_

> Variable, dimension and attribute names should begin with a letter and be composed of letters, digits, and underscores.

That change was made by @davidhassell in https://github.com/cf-convention/cf-conventions/commit/2a44cccf65fa0b1dedffa41ebafe4e03d3d2fa6b and https://github.com/cf-convention/cf-conventions/commit/c3fa6fd5ac220a707250b7523bfbd6c1cdb103d6. Do you remember why this change was made, David?

According to principle 9 of sect 1.2, we shouldn't revert to making it a requirement:

> Because many datasets remain in use for a long time after production, it is desirable that metadata written according to previous versions of the convention should also be compliant with and have the same interpretation under later versions.

Therefore I propose that we change the first sentence of 2.3 to read

> It is recommended that variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

which makes it consistent with the present conformance document. I believe that all those who've contributed recently think that this is what the text should mean. Are you content with making this change?

**Comment by davidhassell** [comment]
Hello @JonathanGregory,

> That change was made by @davidhassell in https://github.com/cf-convention/cf-conventions/commit/2a44cccf65fa0b1dedffa41ebafe4e03d3d2fa6b and https://github.com/cf-convention/cf-conventions/commit/c3fa6fd5ac220a707250b7523bfbd6c1cdb103d6. Do you remember why this change was made, David?

Those commits were from PR #227 that fixed issue #226 (_Correct the wording in the conformance document section 2.3 "Naming Conventions"_).

Thanks, David

**Comment by JonathanGregory** [comment]
Dear @davidhassell

Thanks. I didn't remember about https://github.com/cf-convention/cf-conventions/issues/226, where we previously decided that "should" was intended mean a recommendation. Since the discussion above shows that it is open to question, I believe that my proposal to change the text in sect 2.3 would be helpful, from

> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

to

> It is recommended that variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

I'm relabelling this issue as a `defect`, meaning that the above change will be adopted three weeks from now (10th June) if no-one disagrees before then.

Best wishes

Jonathan

**Comment by larsbarring** [comment]
Well aware of my ever so often much too "free and relaxed interpretation" of English spelling and grammar, I nevertheless venture to ask if it would be possible to somehow exclude the "should" in the suggested wording:

>   It is recommended that variable, dimension, attribute and group names ~~should~~ begin with a letter and be composed of letters, digits, and underscores.

?

**Comment by davidhassell** [comment]
Hi Lars,

That sounds like a good suggestion. BCP14 says 
```
SHOULD   This word, or the adjective "RECOMMENDED", mean that there
may exist valid reasons in particular circumstances to ignore a
particular item, but the full implications must be understood and
carefully weighed before choosing a different course.
```
so providing both words (... recommended ... should ...) doesn't add anything beyond using just one of them.

**Comment by JonathanGregory** [comment]
It's true that "should" doesn't convey any information, given "recommended" for clarity. It would be OK in English to say

> It is recommended that variable, dimension, attribute and group names begin with a letter and be composed of letters, digits, and underscores.

where _begin_ is a subjunctive (a vestigial feature of English grammar). That's not such a common construction though. Maybe some readers might find it obscure? What do you think of

> It is recommended _for_ variable, dimension, attribute and group names _to begin_ with a letter and be composed of letters, digits, and underscores.

or

> Variable, dimension, attribute and group names _are recommended to begin_ with a letter and be composed of letters, digits, and underscores.

**Comment by taylor13** [comment]
I too recommend that we should avoid both "should" and "recommend" in the same sentence.  :)  .  Personally, I prefer the first of the 3 options appearing in the previous post (with the subjunctive construct).  I don't find it confusing.  Perhaps I'm just a vestige of a disappearing generation, so as a second choice I might slightly prefer "are recommended to begin", but that seems a bit awkward to me.

**Comment by MTG-Formats** [comment]
Is there some reference that can be added where users can read the disadvantages/problems they may have  if they don't follow the recommendations?

**Comment by JonathanGregory** [comment]
Four weeks have passed without objection to the proposed remedy for the defect. Therefore we've agreed to make the change, and I've prepared [pull request 526](https://github.com/cf-convention/cf-conventions/pull/526) to implement it. The PR replaces the existing sentence in 2.3

> Variable, dimension, attribute and group names should begin with a letter and be composed of letters, digits, and underscores.

with the wording preferred by @larsbarring and Karl @taylor13 

> It is recommended that variable, dimension, attribute and group names begin with a letter and be composed of letters, digits, and underscores.

to indicate that this is not a requirement, but a recommendation, as shown by the conformance document. Please could someone check and merge this PR e.g. @larsbarring or @davidhassell?

In addition, I am labelling this issue for consideration as a FAQ, in view of the question from Tim @MTG-Formats "Is there some reference that can be added where users can read the disadvantages/problems they may have if they don't follow the recommendations?" It seems to me that if someone has time it would be useful to summarise the early discussion about the advantages of sticking to the convention in the FAQ, or at least we could refer to this issue as a reference from the FAQ.

Thanks to all for contributions to this issue and to @Dave-Allured for raising it.

**Comment by larsbarring** [comment]
@JonathanGregory I have just approved and merged the PR. But it just struck me that the label `change agreed` is both _correct_ as we did agree on some changes, and _incorrect_ as the changes we agreed on are rather the opposite of the initial suggestion. I wonder whether it would be prudent/relevant/informative to actually use both labels, `change agreed` _and_ `agreement not to change` ? This may seem as confusing for a reader, but it  may seem even more confusing if the reader is searching the conventions text for CF support for a wide range of Unicode characters.

**Comment by JonathanGregory** [comment]
Dear @larsbarring

Thanks for merging the PR. I see your point about `change agreed`. It was intended to be the opposite of `agreement not to change`. They should be mutually exclusive. The aim is to indicate why the issue was closed. To avoid this possible confusion, I suggest we should rename `change agreed` to something clearer, which indicates that _some_ change was agreed, although not necessarily what was originally proposed. That's often the case, of course. The renamed label would appear in all the places where `change agreed` currently appears i.e. it's the same identity, just different text.

Best wishes

Jonathan

**Comment by JonathanGregory** [comment]
What do you think of `convention was changed`? Does it avoid the problem you raised?
